File: Engine/src/com/shinkson47/OPEX/backend/toolbox/OPEXBackEndUtils.java
package com.shinkson47.OPEX.backend.toolbox;

import java.io.ByteArrayInputStream;
import java.io.InputStream;
import java.io.UnsupportedEncodingException;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

import com.shinkson47.OPEX.backend.errormanagement.EMSHelper;
import com.shinkson47.OPEX.backend.errormanagement.exceptions.OPEXStaticException;
import com.shinkson47.OPEX.backend.runtime.threading.OPEXGame;

public class OPEXBackEndUtils {

	private OPEXBackEndUtils() throws OPEXStaticException {
		throw new OPEXStaticException("Back End Tools is not instantiable!");
	}

	public static void getAndInvokeMethod(Class<?> cls, String MethodName, Class<?> ParameterTypes, Object parameters,
			Object caller) throws IllegalAccessException, IllegalArgumentException, InvocationTargetException,
			NoSuchMethodException, SecurityException {
		invokeMethod(cls.getMethod(MethodName, ParameterTypes), caller, parameters);
	}

	public static void invokeMethod(Method mtd, Object caller, Object parameters)
			throws IllegalAccessException, IllegalArgumentException, InvocationTargetException {
		mtd.invoke(caller, parameters);
	}

	public static void ParseToSystemIn(String data){
		InputStream old = System.in;
		try {
			InputStream testInput = new ByteArrayInputStream( data.getBytes("UTF-8") );
			System.setIn(testInput);
		} catch (UnsupportedEncodingException e) {
			EMSHelper.handleException(e);
		} finally {
			System.setIn(old);
		}
	}


}

File: Engine/src/com/shinkson47/opex/backend/toolbox/OPEXBackEndUtils.kt
package com.shinkson47.opex.backend.toolbox

import com.shinkson47.opex.backend.errormanagement.EMSHelper
import java.io.ByteArrayInputStream
import java.io.InputStream
import java.io.UnsupportedEncodingException
import java.lang.reflect.InvocationTargetException
import java.lang.reflect.Method

/**
 * Container for general tool methods and functions
 * //TODO these should have thier own classes.
 */

@Throws(IllegalAccessException::class, IllegalArgumentException::class, InvocationTargetException::class, NoSuchMethodException::class, SecurityException::class)
fun getAndInvokeMethod(cls: Class<*>, MethodName: String?, ParameterTypes: Class<*>?, parameters: Any?, caller: Any?) {
    invokeMethod(cls.getMethod(MethodName, ParameterTypes), caller, parameters)
        }

@Throws(IllegalAccessException::class, IllegalArgumentException::class, InvocationTargetException::class)
fun invokeMethod(mtd: Method, caller: Any?, parameters: Any?) {
    mtd.invoke(caller, parameters)
}

fun ParseToSystemIn(data: String) {
    val old = System.`in`
    try {
        val testInput: InputStream = ByteArrayInputStream(data.toByteArray(charset("UTF-8")))
        System.setIn(testInput)
    } catch (e: UnsupportedEncodingException) {
        EMSHelper.handleException(e)
    } finally {
        System.setIn(old)
    }
}

